pkg/kubelet/custommetrics/impl.go

```diff
@@ -0,0 +1,149 @@
+		if err != nil {
+			return nil, err
+		}
+		return customMetricsMount, nil
+	}
+	return nil, nil
+}
+
+func makeCustomMetricsMount(container *api.Container, podContainerDir, podIP, definition string, maxCustomMetricsPerPod int) (*kubecontainer.Mount, error) {
+	endpoints, err := getCustomMetricsEndpointsForContainer(container, definition, maxCustomMetricsPerPod)
+	if err != nil {
+		return nil, err
+	}
+	if len(endpoints) == 0 {
+		return nil, nil
+	}
+
+	if err := os.MkdirAll(podContainerDir, 0750); err != nil {
+		return nil, fmt.Errorf("error creating pod-container directory: %v", err)
+	}
+	hostsFilePath := path.Join(podContainerDir, "cadvisor-custom-metrics-prometheus.json")
+	if err := ensureCustomMetricsFile(hostsFilePath, podIP, endpoints); err != nil {
+		return nil, err
+	}
+	return &kubecontainer.Mount{
+		Name:          "k8s-custom-metrics",
+		ContainerPath: CustomMetricsDefinitionContainerPath,
+		HostPath:      hostsFilePath,
+		ReadOnly:      false,
+	}, nil
+}
+
+func getCustomMetricsEndpointsForContainer(container *api.Container, definition string, maxCustomMetricsPerPod int) ([]CustomMetricsEndpoint, error) {
+	var endpoints []CustomMetricsEndpoint
+	err := json.Unmarshal([]byte(definition), &endpoints)
+	if err != nil {
+		return nil, fmt.Errorf("error while parsing custom metrics definition '%s': %v", definition, err)
+	}
+
+	numCustomMetrics := 0
+	for _, endpoint := range endpoints {
+		if len(endpoint.Names) == 0 {
+			return nil, fmt.Errorf("custom metrics endpoint definition cannot contain empty names: %s", endpoint)
+		}
+		numCustomMetrics += len(endpoint.Names)
+	}
+	if numCustomMetrics > maxCustomMetricsPerPod {
+		return nil, fmt.Errorf("too many custom metrics per pod, only %d allowed: %v", maxCustomMetricsPerPod, endpoints)
+	}
+
+	var matching []CustomMetricsEndpoint
+	for _, port := range container.Ports {
+		for _, endpoint := range endpoints {
+			if port.ContainerPort == endpoint.Port {
+				matching = append(matching, endpoint)
+			}
+		}
+	}
+	return matching, nil
+}
+
+func ensureCustomMetricsFile(fileName, hostIP string, endpoints []CustomMetricsEndpoint) error {
+	if len(endpoints) > 1 {
+		return fmt.Errorf("too many endpoints (only one endpoint per container is allowed)")
+	}
+	endpoint := endpoints[0]
+	if len(endpoint.Names) == 0 {
+		return fmt.Errorf("names of custom metrics to collect cannot be empty")
+	}
+
+	if _, err := os.Stat(fileName); os.IsExist(err) {
+		glog.V(4).Infof("kubernetes-managed metrics file exits. Will not be recreated: %q", fileName)
+		return nil
+	}
+	var buffer bytes.Buffer
+	buffer.WriteString("{\n")
+	buffer.WriteString(fmt.Sprintf("  \"endpoint\" : \"http://%s:%d%s\",\n", hostIP, endpoint.Port, endpoint.Path))
+	buffer.WriteString("  \"metrics_config\" : [\n")
+	for i, name := range endpoint.Names {
+		comma := ","
+		if i == len(endpoint.Names)-1 { // Do not print comma after the last name.
```

Comment:
The issue is that you need to, at the very least, quote everything first.  Additionally, if you want the indentation, that gets a bit weird looking with strings.Join as well.